rust/src/cxx_bridge_gobject.rs

```diff
@@ -0,0 +1,54 @@
+//! Wrappers that bridge the world of cxx.rs and GObject-introspect based crate bindings.
+//! See https://github.com/dtolnay/cxx/issues/544
+//! While cxx.rs supports including externally-bound types (like ostree::Repo),
+//! two things make this more complicated.  First, cxx.rs requires implementing
+//! a trait, and due to the orphan rule we can't implement foreign traits on
+//! foreign types.  Second, what we *actually* want on the Rust side isn't
+//! the *_sys type (e.g. ostree_sys::OstreeRepo) but the `ostree::Repo` type.
+//! So for now, we define a `FFIGObjectWrapper` trait that helps with this.
+//! In the future though hopefully cxx.rs improves this situation.
+
+use cxx::{type_id, ExternType};
+
+pub trait FFIGObjectWrapper {
+    type Wrapper;
+
+    fn gobj_wrap(&mut self) -> Self::Wrapper;
+}
+
+macro_rules! wrap {
+    ($w:ident, $bound:path) => {
+        impl FFIGObjectWrapper for $w {
+            type Wrapper = $bound;
+            fn gobj_wrap(&mut self) -> Self::Wrapper {
+                unsafe { glib::translate::from_glib_none(&mut self.0 as *mut _) }
+            }
+        }
+    };
+}
+#[repr(transparent)]
+pub struct FFIOstreeSysroot(ostree_sys::OstreeSysroot);
```

Comment:
So I was looking at improving the `wrap!` macro to also generate this stuff, I got as far as this:

```
diff --git a/Cargo.lock b/Cargo.lock
index 990bc81c..72add85e 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -771,6 +771,12 @@ dependencies = [
  "system-deps",
 ]
 
+[[package]]
+name = "paste"
+version = "1.0.4"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "c5d65c4d95931acda4498f675e332fcbdc9a06705cd07086c510e9b6009cd1c1"
+
 [[package]]
 name = "pin-project"
 version = "0.4.27"
@@ -1002,6 +1008,7 @@ dependencies = [
  "openat-ext",
  "ostree",
  "ostree-sys",
+ "paste",
  "rand",
  "rayon",
  "serde",
diff --git a/Cargo.toml b/Cargo.toml
index cf8a0b72..b4b23570 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -6,6 +6,7 @@ edition = "2018"
 
 [dependencies]
 anyhow = "1.0.35"
+paste = "1.0"
 serde = "1.0.118"
 serde_derive = "1.0.118"
 serde_json = "1.0.60"
diff --git a/rust/src/cxx_bridge_gobject.rs b/rust/src/cxx_bridge_gobject.rs
index d595017c..a2c5aa69 100644
--- a/rust/src/cxx_bridge_gobject.rs
+++ b/rust/src/cxx_bridge_gobject.rs
@@ -9,6 +9,7 @@
 //! In the future though hopefully cxx.rs improves this situation.
 
 use cxx::{type_id, ExternType};
+use paste::paste;
 
 pub trait FFIGObjectWrapper {
     type Wrapper;
@@ -16,7 +17,7 @@ pub trait FFIGObjectWrapper {
     fn gobj_wrap(&mut self) -> Self::Wrapper;
 }
 
-macro_rules! wrap {
+macro_rules! impl_wrap {
     ($w:ident, $bound:path) => {
         impl FFIGObjectWrapper for $w {
             type Wrapper = $bound;
@@ -26,14 +27,23 @@ macro_rules! wrap {
         }
     };
 }
-#[repr(transparent)]
-pub struct FFIOstreeSysroot(ostree_sys::OstreeSysroot);
 
-unsafe impl ExternType for FFIOstreeSysroot {
-    type Id = type_id!("rpmostreecxx::OstreeSysroot");
-    type Kind = cxx::kind::Trivial;
+macro_rules! bind_ostree_obj {
+    ($w:ident) => {
+        paste! {
+            #[repr(transparent)]
+            pub struct [<FFIOstree $w>](ostree_sys::[<Ostree $w>]);
+            
+            unsafe impl ExternType for [<FFIOstree $w>] {
+                type Id = type_id!(stringify!([< rpmostreecxx::Ostree $w>]));
+                type Kind = cxx::kind::Trivial;
+            }
+            impl_wrap!([<FFIOstree $w>], ostree::$w);
+        }
+    };
 }
-wrap!(FFIOstreeSysroot, ostree::Sysroot);
+
+bind_ostree_obj!(Sysroot);
 
 #[repr(transparent)]
 pub struct FFIOstreeRepo(ostree_sys::OstreeRepo);
@@ -42,7 +52,7 @@ unsafe impl ExternType for FFIOstreeRepo {
     type Id = type_id!("rpmostreecxx::OstreeRepo");
     type Kind = cxx::kind::Trivial;
 }
-wrap!(FFIOstreeRepo, ostree::Repo);
+impl_wrap!(FFIOstreeRepo, ostree::Repo);
 
 #[repr(transparent)]
 pub struct FFIOstreeDeployment(ostree_sys::OstreeDeployment);
@@ -51,4 +61,4 @@ unsafe impl ExternType for FFIOstreeDeployment {
     type Id = type_id!("rpmostreecxx::OstreeDeployment");
     type Kind = cxx::kind::Trivial;
 }
-wrap!(FFIOstreeDeployment, ostree::Deployment);
+impl_wrap!(FFIOstreeDeployment, ostree::Deployment);
diff --git a/src/libpriv/rpmostree-cxxrs-prelude.h b/src/libpriv/rpmostree-cxxrs-prelude.h
index 7ce882bc..c9a54b7c 100644
--- a/src/libpriv/rpmostree-cxxrs-prelude.h
+++ b/src/libpriv/rpmostree-cxxrs-prelude.h
@@ -26,5 +26,4 @@ namespace rpmostreecxx {
     typedef ::OstreeSysroot OstreeSysroot;
     typedef ::OstreeRepo OstreeRepo;
     typedef ::OstreeDeployment OstreeDeployment;
-    typedef char* GString;
 }
```

but I can't quite figure out how to make the `type_id!` part work...I think it only accepts a string but could accept a generated path too maybe?

Reply:
Seems like you probably meant `type_id!(rpmostreecxx::[<Ostree $w>])`.

Reply:
Ah yes, thank you!